### PR TITLE
[mod_menu] too much duplicate queries in menu.php

### PR DESCRIPTION
**Pull request by alikon**
current staging
enable debug
Administrator->Control panel
see the database queries tab look at duplicates found

### Testing Instructions
admin menu works as expected

### Expected result
less duplicated query
screenshot from 2017-10-07 08-52-51

### Actual result
screenshot from 2017-10-07 08-49-38

### CODE REVIEW

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on eebc59c71ad7aea5ff75b4ae4edf41e8b02e0222<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Comment by infograf768** [comment]
This code has just been merged as com_fields menus were displaying when fields is disabled in Extensions=>Manage
See:
https://github.com/joomla/joomla-cms/pull/18134

**Comment by infograf768** [comment]
I have tested this item :red_circle: unsuccessfully on eebc59c71ad7aea5ff75b4ae4edf41e8b02e0222<br /><br />See https://github.com/joomla/joomla-cms/pull/18134<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Comment by alikon** [comment]
@infograf768  anyway not in that way #18134

**Comment by infograf768** [comment]
Suggested a solution there.
```
			// Exclude item if com_fields component is not installed or disabled
			if ($item->element == 'com_fields' && (!JComponentHelper::isInstalled($item->element) || !JComponentHelper::isEnabled($item->element)))
			{
				continue;
			}
```

**Comment by alikon** [comment]
only for `com_fields` sound more reasonable

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on a074b0944cf5de28e47ce88240f98b2003a9bb70<br /><br />Yep. :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Comment by izharaazmi** [comment]
https://github.com/joomla/joomla-cms/pull/18134#issuecomment-334918725. Do we really want to show other components link if they are not installed or disabled?

**Comment by alikon** [comment]
i guess 
https://github.com/alikon/joomla-cms/blob/a074b0944cf5de28e47ce88240f98b2003a9bb70/administrator/modules/mod_menu/menu.php#L95

**Comment by infograf768** [comment]
I tested with other components without issues. com_fields is very specific.

**Comment by infograf768** [comment]
tested further.
`if ($item->element && !JComponentHelper::isEnabled($item->element))`
is not the one adding queries

it is
`!JComponentHelper::isInstalled($item->element)` which is the culprit.

Basically only a component which does not exist in the _extensions table but still has entries in the _menu table would be concerned.

FYI `JComponentHelper::isInstalled` is never used in core.

**Comment by izharaazmi** [comment]
@infograf768 As I can see `\Joomla\CMS\Component\ComponentHelper::isInstalled` return the count of records matching the component name passed. I wonder whether it is possible that the value is anything other than `0` or `1`?

If not then IMHO, we can simply proxy this method to `isEnabled` and deprecated this for J4.

**Comment by infograf768** [comment]
@izharaazmi @alikon 

> If not then IMHO, we can simply proxy this method to isEnabled and deprecated this for J4.

I would not do that as it would not be B/C in case a 3rd party is using the method.

But in our case here, I just would not use it at all and change this PR to use only
`if ($item->element && !JComponentHelper::isEnabled($item->element))`

@mbabker 
What do you think?

**Comment by izharaazmi** [comment]
>  I just would not use it at all and change this PR to use only ...

Fair enough for me.

**Comment by infograf768** [comment]
@alikon 
Can you modify this PR?

**Comment by alikon** [comment]
if `isEnabled() =yes/no`  i will assume that `isInstalled() = true` :laughing:

**Comment by infograf768** [comment]
> if isEnabled() =yes/no i will assume that isInstalled() 

Yep. Can you modify PR?

**Comment by alikon** [comment]
@infograf768  so no more exception for `com_fields` ?

**Comment by infograf768** [comment]
no need as you can test

**Comment by infograf768** [comment]
You still need imho to add
`if ($item->element`

**Comment by alikon** [comment]
please double check  cause i'm unable to test it right now :stuck_out_tongue_winking_eye:

File: administrator/modules/mod_menu/menu.php [hunk 1]
```diff
@@ -275,8 +275,8 @@ protected function preprocess($items)
 				}
 			}
 
-			// Exclude item if the component is not installed or disabled
-			if ($item->element && (!JComponentHelper::isInstalled($item->element) || !JComponentHelper::isEnabled($item->element)))
+			// Exclude item if is not enabled
+			if ($item->element && !JComponentHelper::isEnabled($item->element))
```
Comment (izharaazmi) [inline administrator/modules/mod_menu/menu.php:279]: I think we should strict check. `JComponentHelper::isEnabled($item->element, true)`

Comment (izharaazmi) [reply administrator/modules/mod_menu/menu.php:279]: Sorry, its implicit. Please ignore

Comment (alikon) [reply administrator/modules/mod_menu/menu.php:279]: np more eyes is always better

**Comment by infograf768** [comment]
Fine now.
screen shot 2017-10-11 at 11 45 13

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on f92c623ae4ed74a53adaec0219325aa589e5ee72<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Comment by mbabker** [comment]
> As I can see \Joomla\CMS\Component\ComponentHelper::isInstalled return the count of records matching the component name passed. I wonder whether it is possible that the value is anything other than 0 or 1?

The method should only return 0 or 1.  If it doesn't, then someone has a corrupt database because you can't install two extensions with the same component name.

**Comment by izharaazmi** [comment]
@infograf768 @mbabker @alikon Please also see #18308.

PS: This PR is still applicable. My PR does not replace it.

**Comment by infograf768** [comment]
@izharaazmi 
Please nevertheless set this one to test OK

**Comment by izharaazmi** [comment]
I have tested this item :white_check_mark: successfully on f92c623ae4ed74a53adaec0219325aa589e5ee72<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Comment by infograf768** [comment]
RTC. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18264">issues.joomla.org/tracker/joomla-cms/18264</a>.</sub>

**Pull request merged**